File: apps/node/src/chapter08/survived-data.ts
import { readFileSync } from "node:fs";
import { parse } from "csv-parse/sync";

export const FEATURES = [
  "Pclass",
  "Sex",
  "Age",
  "SibSp",
  "Parch",
  "Fare",
  "Embarked",
] as const;
export const TARGET = "Survived";

/** 乗客 1 人分の特徴量 */
export interface Passenger {
  Pclass: number;
  Sex: string;
  Age: number | null;
  SibSp: number;
  Parch: number;
  Fare: number;
  Embarked: string | null;
}

export interface SurvivedRow extends Passenger {
  PassengerId: number;
  Survived: number;
  Ticket: string;
  Cabin: string | null;
}

const STRING_COLUMNS: readonly string[] = [
  "Sex",
  "Ticket",
  "Cabin",
  "Embarked",
];

export function loadSurvived(csvFile: string): SurvivedRow[] {
  return parse<SurvivedRow>(readFileSync(csvFile), {
    bom: true,
    columns: true,
    cast: (value, context) => {
      if (context.header) {
        return value;
      }
      if (value === "") {
        return null;
      }
      return STRING_COLUMNS.includes(String(context.column))
        ? value
        : Number(value);
    },
  });
}

export function splitFeaturesAndTarget(rows: readonly SurvivedRow[]): {
  x: Passenger[];
  t: number[];
} {
  return {
    x: rows.map(
      ({
        PassengerId: _id,
        Survived: _survived,
        Ticket: _ticket,
        Cabin: _cabin,
        ...features
      }) => features,
    ),
    t: rows.map((row) => row[TARGET]),
  };
}

File: apps/java/src/main/java/chapter08/SurvivedData.java
package chapter08;

import chapter02.Row;
import chapter02.Table;
import java.util.List;

/** Survived.csv の特徴量の列と正解ラベルの列。 */
public final class SurvivedData {
  /** モデルに渡す特徴量の列 */
  public static final List<String> FEATURES =
      List.of("Pclass", "Sex", "Age", "SibSp", "Parch", "Fare", "Embarked");

  /** 正解ラベルの列（1 が生存、0 が死亡） */
  public static final String TARGET = "Survived";

  private SurvivedData() {}

  /** 行を、特徴量の列だけを並べた表にする。行がほかの列を持っていても使わない。 */
  public static Table features(List<Row> rows) {
    return new Table(FEATURES, rows);
  }

  /** 行の Survived 列を、整数の正解ラベルにする。 */
  public static List<Integer> target(List<Row> rows) {
    return rows.stream().map(row -> Integer.parseInt(row.text(TARGET))).toList();
  }
}
